conda-store-server/tests/test_db_api.py

```diff
@@ -39,35 +39,104 @@ def test_namespace_crud(db):
+    assert len(api.list_namespaces(db).all()) == 4
+
+    # Creates role mappings
+    api.create_namespace_role(db, name=namespace_name, other=other_namespace_name1, role="admin")
+    api.create_namespace_role(db, name=namespace_name, other=other_namespace_name2, role="admin")
+    api.create_namespace_role(db, name=namespace_name, other=other_namespace_name3, role="viewer")
+    db.commit()
 
-    # create a Role Mapping, with a failing entity
-    with pytest.raises(Exception):
-        NamespaceRoleMapping(
-            namespace=namespace, namespace_id=namespace.id, entity="invalid_entity_name"
-        )
+    # Attempts to create a role mapping with an invalid role
+    with pytest.raises(ValueError, match=r"invalid role=invalid-role"):
+        api.create_namespace_role(db, name=namespace_name, other=other_namespace_name3, role="invalid-role")
+        db.commit()
+
+    # Attempts to create a role mapping violating the uniqueness constraint
+    with pytest.raises(
+            Exception,
+            match=(r"UNIQUE constraint failed: "
+                   r"namespace_role_mapping.namespace_id, "
+                   r"namespace_role_mapping.other_namespace_id")):
+        # Runs in a nested transaction since a constraint violation will cause a rollback
+        with db.begin_nested():
+            api.create_namespace_role(db, name=namespace_name, other=other_namespace_name2, role="developer")
+            db.commit()
+
+    # Updates a role mapping
+    api.update_namespace_role(db, name=namespace_name, other=other_namespace_name2, role="developer")
+    db.commit()
 
-    # Creates role mappings with valid entity names
-    NamespaceRoleMapping(namespace=namespace, namespace_id=namespace.id, entity="org/*")
+    # Gets all role mappings
+    roles = api.get_namespace_roles(db, namespace_name)
+    db.commit()
+    assert len(roles) == 3
 
-    NamespaceRoleMapping(
-        namespace=namespace, namespace_id=namespace.id, entity="*/team"
-    )
+    assert roles[0].id == 1
+    assert roles[0].namespace == namespace_name
+    assert roles[0].other_namespace == other_namespace_name1
+    assert roles[0].role == 'admin'
 
-    NamespaceRoleMapping(
-        namespace=namespace, namespace_id=namespace.id, entity="org/team"
-    )
+    assert roles[1].id == 2
+    assert roles[1].namespace == namespace_name
+    assert roles[1].other_namespace == other_namespace_name2
+    assert roles[1].role == 'developer'
+
+    assert roles[2].id == 3
+    assert roles[2].namespace == namespace_name
+    assert roles[2].other_namespace == other_namespace_name3
+    assert roles[2].role == 'viewer'
 
-    NamespaceRoleMapping(namespace=namespace, namespace_id=namespace.id, entity="*/*")
+    # Gets other role mappings
+    roles = api.get_other_namespace_roles(db, other_namespace_name1)
+    db.commit()
+    assert len(roles) == 1
+    roles = api.get_other_namespace_roles(db, namespace_name)
+    db.commit()
+    assert len(roles) == 0
```

Comment:
We also test the `other` variant of the get function, which is used in authentication. One namespace is part of the original three mappings, which we've just added, so 1 entry is returned as expected (it has access to one namespace). The namespace that we give access to we didn't insert anywhere, so it has no access to any other namespaces.

Reply:
Perhaps I should check the other values here, as above, not just the lenghts.

Reply:
In general, relying on having a pure db and checking lengths has led to flaky tests for me in the past, especially if you ever try to run tests in parallel. I always prefer to check for concrete records than for lengths.